ddd/ddd-05-mem-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,250 @@
+    means "as small as actually possible". While adaptation of this kind of
+    application (demo 5) will require more work than the other kinds of
+    application (as demonstrated in demos 1 through 4) we can at least provide
+    an interface which applications will find familiar and which lacks any
+    excessive impedence discontinuities with their own code.
+
+  - There already exists BIO_dgram, which is used with DTLS. This causes
+    BIO_read/BIO_write to work similarly to recv/send. This works on
+    a file descriptor; there is no analagous BIO_dgram_mem.
+    Such a BIO could be created, which would model datagram semantics
+    for read/write operations (i.e., a sequence of BIO_write calls
+    with given lengths results in an exactly matching sequence of
+    BIO_read calls with the same lengths returned).
+
+  - There are some caveats for this idea. In order for this to work:
+
+    - applications must ensure there is a guaranteed 1:1 correspondance
+      between send/recv and BIO_read/BIO_write calls; and
+
+    - the buffer passed to BIO_read() and recv() must always be big
+      enough to receive the largest possible packet, else silent
+      truncation occurs.
+
+    The latter is an unfortunate design issue with the Berkeley sockets API
+    for SOCK_DGRAM, which is replicated here in BIO_read which simply serves
+    as a wrapper for it. The caller thus has to know the worst-case UDP
+    datagram payload size, but knowing this requires knowledge of the
+    network or at least the configuration of local network interfaces.
+
+    (Random factoid: this design issue once in the past led me to determine
+    the maximum size of an incoming UDP datagram as
+
+      max(64 KiB, max(MTU of every local interface))
+
+    -- reason being that an IPv6 fragmentation header has a 16-bit length field,
+    so an IP packet greater than 64 KiB (using the IPv6 jumbogram header) can
+    only be sent if you have an interface with a native MTU exceeding it.
+    Another option is to use MSG_PEEK | MSG_TRUNC on systems which support it
+    but this requires calling recv twice per packet.)
+
+  - For the buffer size issue, our options for supporting existing applications
+    include:
+
+    - Using a maximum MTU of 1472 by default; then any application using
+      old BIO_read/BIO_write calls will work so long as it uses a buffer
+      of at least 1472 bytes (and there is a 1:1 correspondance of BIO calls and
+      syscalls).
+
+    - Applications which can support larger MTUs (e.g. a network with
+      jumbo frames) could indicate support via some method TBD, such as
+      capability flags (see below).
+
+    In this model, for the purposes of compatibility and minimising changes:
+
+    - BIO_ctrl_get_write_guarantee (BIO_C_GET_WRITE_GUARANTEE) would always
+      return either 0 or a value >=1472. (This is a change from BIO_dgram,
+      which does not implement BIO_C_GET_WRITE_GUARANTEE.)
```

Comment:
Not sure I understand the need for this?

Reply:
My motivation for this is that if you return a value `x` below 1472, it will cause problems with application code that does something like `BIO_write(buf, min(..., BIO_ctrl_get_write_guarantee()))`.

Essentially the idea here is that the BIO_dgram_mem can either accept another dgram or it can't. It never "partially" accepts dgrams written to it, so it makes no sense for it to claim it can accept dgrams but only if they're smaller than the MTU.

Reply:
BIO_dgram has `BIO_CTRL_DGRAM_GET_MTU`. Isn't this effectively the same?

Reply:
The BIO_dgram_mem will be passed to the SSL object, and we can certainly expect the SSL object to make use of the MTU set on the dgram_mem. Since the application is responsible for setting the MTU value on the dgram_mem, it doesn't really make sense for it to query it from the same dgram_mem.

This is more about minimising surprises on the application side.

Reply:
Somehow I'm still missing something here...